File: legacy/src/GitOpen.Domain/Commits/CommitSha.cs
namespace GitOpen.Domain.Commits;

public readonly record struct CommitSha
{
    public string Value { get; }

    public CommitSha(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ArgumentException("CommitSha cannot be empty", nameof(value));
        if (value.Length is < 4 or > 40)
            throw new ArgumentException("CommitSha must be 4..40 hex chars", nameof(value));
        Value = value.ToLowerInvariant();
    }

#pragma warning disable CA1720 // identifier contains type name
    public string Short(int length = 7) =>
        Value.Length <= length ? Value : Value[..length];
#pragma warning restore CA1720

    public override string ToString() => Value;
}

File: lib/domain/commits/commit_sha.dart
import 'package:equatable/equatable.dart';

final class CommitSha extends Equatable {

  CommitSha(String input) : value = _normalize(input);
  final String value;

  static String _normalize(String input) {
    if (input.trim().isEmpty) {
      throw ArgumentError('CommitSha cannot be empty');
    }
    if (input.length < 4 || input.length > 40) {
      throw ArgumentError('CommitSha must be 4..40 hex chars');
    }
    return input.toLowerCase();
  }

  String short([int length = 7]) =>
      value.length <= length ? value : value.substring(0, length);

  @override
  String toString() => value;

  @override
  List<Object?> get props => [value];
}
